modules/phase_field/include/utils/TensorTempl.h

```diff
@@ -0,0 +1,361 @@
+  friend TensorTempl & operator*(const T & lhs, TensorTempl<T> & rhs) { return rhs * lhs; }
+  TensorTempl &
+  contractMult(const TensorTempl & lhs, const TensorTempl & rhs, unsigned int contractor);
+  void recurseMult(T & result_term,
+                   unsigned int index,
+                   unsigned int divider,
+                   std::vector<long unsigned int> center,
+                   unsigned int current,
+                   std::vector<unsigned int> current_vec,
+                   const TensorTempl & lhs,
+                   const TensorTempl & rhs);
+
+  friend std::ostream & operator<<(std::ostream & os, const TensorTempl<T> & tens)
+  {
+    os << "Shape : {";
+    for (unsigned int i = 0; i < tens._shape.size(); ++i)
+    {
+      os << tens._shape[i];
+      if (i != tens._shape.size() - 1)
+        os << ',';
+    }
+    os << "}" << std::endl;
+    os << "Accessor : {";
+    for (unsigned int i = 0; i < tens._access_data.size(); ++i)
+    {
+      os << tens._access_data[i];
+      if (i != tens._access_data.size() - 1)
+        os << ',';
+    }
+    os << "}" << std::endl;
+    os << "Data : {";
+    for (unsigned int i = 0; i < tens._data.size(); ++i)
+    {
+      os << tens._data[i];
+      if (i != tens._data.size() - 1)
+        os << ',';
+    }
+    os << "}" << std::endl;
+    return os;
+  }
+
+  /// works for order 1 and 2 only
+  void transpose();
+  void transpose(unsigned int first_index, unsigned int second_index);
+
+  void setAccessor();
+
+  void applyFunction(void (*function)(T & t));
+
+protected:
+  std::vector<T> _data;
+  std::vector<std::size_t> _shape;
+  std::vector<unsigned int> _access_data;
+
+private:
+  // initializer list processing
+  template <typename U>
+  void initialize(std::initializer_list<U> u, std::size_t _depth);
+  void initialize(std::initializer_list<T> t, std::size_t _depth);
+  void checkDimensions(std::size_t size, std::size_t _depth);
+
+  // helper index
+  std::size_t _i;
+};
+
+// operators
+template <typename T>
+T
+TensorTempl<T>::operator()(unsigned int index, unsigned int index_rest...)
+{
+  return (*this)(std::vector<unsigned int>(1, index), index_rest);
+}
+
+template <typename T>
+T
+TensorTempl<T>::
+operator()(std::vector<unsigned int> index_caught, unsigned int index, unsigned int index_rest...)
+{
+  index_caught.push_back(index);
+  return (*this)(index_caught, index_rest);
+}
+
+template <typename T>
+T
+TensorTempl<T>::operator()(std::vector<unsigned int> index_caught, unsigned int index)
+{
+  index_caught.push_back(index);
+  unsigned int location = 0;
+  for (unsigned int i = 0; i < _access_data.size(); ++i)
+  {
+    if (index_caught[i] >= _shape[i])
+      mooseError("Incorrect indices for accessing tensor.");
+    location += _access_data[i] * index_caught[i];
+  }
+  return _data[location];
+}
+
+template <typename T>
+T
+TensorTempl<T>::operator()(unsigned int index)
+{
+  std::vector<unsigned int> index_caught(1, index);
```

Comment:
I don't get that. This creates a single element vector that contains just `index`. Why bother building this?

Reply:
Yes, I need to get rid of that. It was part of the learning process for variadic function arguments. This appears in EBTerm as well when doing substitution.